crates/core/executor/src/profiler.rs

```diff
@@ -0,0 +1,213 @@
+use gecko_profile::{Frame, ProfileBuilder, StringIndex, ThreadBuilder};
+use goblin::elf::{sym::STT_FUNC, Elf};
+use indicatif::{ProgressBar, ProgressStyle};
+use rustc_demangle::demangle;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ProfilerError {
+    #[error("Failed to read ELF file {}", .0)]
+    Io(#[from] std::io::Error),
+    #[error("Failed to parse ELF file {}", .0)]
+    Elf(#[from] goblin::error::Error),
+    #[error("Failed to serialize samples {}", .0)]
+    Serde(#[from] serde_json::Error),
+}
+
+pub struct Profiler {
+    sample_rate: u64,
+    start_lookup: HashMap<u64, usize>,
+    function_ranges: Vec<(u64, u64, Frame)>,
+
+    function_stack: Vec<Frame>,
+    function_stack_indices: Vec<usize>,
+    function_stack_ranges: Vec<(u64, u64)>,
+    current_function_range: (u64, u64),
+
+    main_idx: Option<StringIndex>,
+    builder: ThreadBuilder,
+    samples: Vec<Sample>,
+}
+
+struct Sample {
+    stack: Vec<Frame>,
+}
+
+impl Profiler {
+    pub(super) fn new(elf_bytes: &[u8], sample_rate: u64) -> Result<Self, ProfilerError> {
+        let elf = Elf::parse(elf_bytes)?;
+
+        let mut start_lookup = HashMap::new();
+        let mut function_ranges = Vec::new();
+        let mut builder = ThreadBuilder::new(1, 0, std::time::Instant::now(), false, false);
+
+        let mut main_idx = None;
+        for sym in &elf.syms {
+            if sym.st_type() == STT_FUNC {
```

Comment:
can you comment what this means -- is it the start of a function?

Reply:
in general i can't follow the logic in here super well, could you add some comments?